toucan_connectors/databricks/databricks_connector.py

```diff
@@ -0,0 +1,129 @@
+from typing import Any, Optional
+
+import backoff
+import pandas as pd
+import pyodbc
+from pydantic import Field, SecretStr, constr
+
+from toucan_connectors.common import ConnectorStatus, pandas_read_sql
+from toucan_connectors.toucan_connector import ToucanConnector, ToucanDataSource
+
+CLUSTER_START_TIMEOUT = 90
+
+
+class DatabricksDataSource(ToucanDataSource):
+    query: constr(min_length=1) = Field(
+        ...,
+        description='You can write a query here',
+        widget='sql',
+    )
+
+
+class DataBricksConnectionError(Exception):
+    """ """
+
+
+class DataBricksUnauthorizedError(Exception):
+    """ """
+
+
+class DatabricksConnector(ToucanConnector):
+    data_source_model: DatabricksDataSource
+    Host: str = Field(
+        ...,
+        description='The listening address of your databricks cluster',
+        placeholder='my-databricks-cluster.cloudproviderdatabricks.net',
+    )
+    Port: int = Field(
+        ..., description='The listening port of your databricks cluster', placeholder=443
+    )
+    HTTPPath: str = Field(
+        ..., description='Databricks compute resources URL', placeholder='sql/protocolv1/o/xxx/yyy'
+    )
+    PWD: SecretStr = Field(
+        None, description='Your personal access token', placeholder='dapixxxxxxxxxxx'
+    )
+    Ansi: bool = False
+    connect_timeout: int = None
+
+    @backoff.on_exception(backoff.expo, pyodbc.Error, max_time=CLUSTER_START_TIMEOUT)
+    def _connect(self) -> Any:
+        """
+        In case of on-demand clusters, the connector might fail to connect.
+        However, the first connection trial starts the cluster so this method will exponentially retry
+        to connect until CLUSTER_START_TIMEOUT is reached.
+        """
+        try:
+            return pyodbc.connect(self._build_connection_string(), autocommit=True, ansi=self.Ansi)
+        except pyodbc.InterfaceError as e:
+            if len(e.args) >= 2 and 'Authentication/authorization error occured' in e.args[1]:
+                raise DataBricksUnauthorizedError('Invalid credentials')
+            else:
+                raise DataBricksConnectionError('Invalid connection params')
+
+    def _build_connection_string(self) -> str:
+        """For constants see: https://docs.databricks.com/integrations/bi/jdbc-odbc-bi.html"""
+        connection_params = {
+            'Driver': '/opt/simba/spark/lib/64/libsparkodbc_sb64.so',  # This path must be correct when installing
+            # ODBC driver
+            'Host': '127.0.0.0.1' if self.Host == 'localhost' else self.Host,
```

Comment:
```suggestion
            'Host': '127.0.0.1' if self.Host == 'localhost' else self.Host,
```

Reply:
And maybe handle the case of `0.0.0.0` ?

Reply:
thx, I guess we can rewrite it to '127.0.0.1'